bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java

```diff
@@ -0,0 +1,474 @@
+                            "Error communicating with Tibber API: " + e.getMessage());
+                }
+            }
+
+        } else if (jsonResponse.contains("error")) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Error in response from Tibber API: " + jsonResponse);
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Unexpected response from Tibber");
+        }
+        return;
+    }
+
+    public void startRefresh(int refresh) {
+        if (pollingJob == null) {
+            pollingJob = scheduler.scheduleWithFixedDelay(() -> {
+                try {
+                    updateRequest();
+                } catch (IOException e) {
+                    logger.warn("IO Exception: ", e);
+                }
+            }, 1, refresh, TimeUnit.MINUTES);
+        }
+        return;
+    }
+
+    public void updateRequest() throws IOException {
+        getURLInput(BASE_URL);
+
+        if ("true".equals(rtEnabled) && !isConnected()) {
+            try {
+                open();
+            } catch (Exception e) {
+                logger.warn("Exception: ", e);
+            }
+        }
+        return;
+    }
+
+    public void updateChannel(String channelID, String channelValue) {
+        if (!channelValue.contains("null")) {
+            if (channelID.contains("consumption") || channelID.contains("Consumption")
+                    || channelID.contains("accumulatedProduction")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.KILOWATT_HOUR));
+            } else if (channelID.contains("power") || channelID.contains("Power")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.WATT));
+            } else if (channelID.contains("voltage")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.VOLT));
+            } else if (channelID.contains("live_current")) {
+                updateState(channelID, new QuantityType<>(new BigDecimal(channelValue), SmartHomeUnits.AMPERE));
+            } else {
+                updateState(channelID, new DecimalType(channelValue));
+            }
+        }
+        return;
+    }
+
+    public void thingStatusChanged(ThingStatusInfo thingStatusInfo) {
+        logger.debug("Thing Status updated to {} for device: {}", thingStatusInfo.getStatus(), getThing().getUID());
+        if (thingStatusInfo.getStatus() != ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Unable to communicate with Tibber API");
+        }
+    }
+
+    @Override
+    public void dispose() {
+        if (pollingJob != null) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
```

Comment:
To deal with the final null warning you need to cache the field in a local variable. The reason for the warning is that the null checker is warning about possible NPE in a concurrent environment. In this case it is preventing a possible NPE if dispose() was called by two separate threads at the same time, this wouldn't ever happen but the null checker doesn't know that. In general we find that the warnings are helpful and allow us to write better code.
```suggestion
        ScheduledFuture<?> pollingJob = this.pollingJob;
        if (pollingJob != null) {
            pollingJob.cancel(true);
            this.pollingJob = null;
        }
```

Reply:
Corrected.